src/daemons/linuxd/src/linuxd/mod.rs

```
@@ -738,3 +407,116 @@ impl LinuxDaemon {
         Ok(message)
     }
 }
+
+//==================================================================================================
+// Internal Helpers (Tokio-based message processing)
+//==================================================================================================
+
+impl LinuxDaemon {
+    async fn get_user_vm_messages(
+        &self,
+        user_vm_connections: &mut HashMap<UserVmIdentifier, UserVmHandle>,
+    ) -> Result<Vec<(UserVmIdentifier, UserVmHandle, Message)>, UserVmIdentifier> {
+        let ids: Vec<UserVmIdentifier> = user_vm_connections.keys().cloned().collect();
+        let mut messages: Vec<(UserVmIdentifier, UserVmHandle, Message)> = Vec::new();
+
+        for uvm_id in ids {
+            let Some(uvm_handle) = user_vm_connections.get(&uvm_id).cloned() else {
+                continue;
+            };
+            let message: Message = match Self::recv(uvm_handle.get_user_vm_reader()).await {
```

As mentioned above, I don't think this function is safe to be used in a `select!` statement. This is because it is not cancellation safe. See the tokio docs on cancellation safety [1].

Intuitively, any competing branches in the `select!` will be blocked on an `await`. Whenever one finishes, **all other futures will be dropped**. This means that any progress that has happened in that branch will be lost. This is fine when you are waiting on a socket or on a queue, but not on `get_user_vm_messages` as this is a stateful method.

For example, if another branch in the `select!` returns before we have iterated over all user vms (but after we have iterated over some), the messages that we have in the `messages` variable will be dropped. This will make us silently drop messages.

I am not sure what is the exact design for this. I was looking around and we could either have a fan-in pattern where we spawn one task per user vm stream, or use a future stream where we yield messages, but, in short, I don't think this method is safe as it is.

[1] https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety

---

Fixed.